File: library/src/main/java/me/proxer/library/api/FixRatingDetailsAdapter.java
package me.proxer.library.api;

import com.squareup.moshi.FromJson;
import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;
import me.proxer.library.entity.info.RatingDetails;

import javax.annotation.Nullable;
import java.io.IOException;

/**
 * @author Ruben Gees
 */
class FixRatingDetailsAdapter {

    private static final String INVALID_ARRAY = "[]";

    private final JsonAdapter<RatingDetails> internalAdapter;

    FixRatingDetailsAdapter() {
        internalAdapter = new Moshi.Builder().build().adapter(RatingDetails.class);
    }

    @FromJson
    @Nullable
    RatingDetails fromJson(final String json) throws IOException {
        if (json.isEmpty() || json.equals(INVALID_ARRAY)) {
            return new RatingDetails(0, 0, 0, 0, 0);
        } else {
            return internalAdapter.fromJson(json);
        }
    }
}

File: library/src/main/kotlin/me/proxer/library/api/FixRatingDetailsAdapter.kt
package me.proxer.library.api

import com.squareup.moshi.FromJson
import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.JsonReader
import me.proxer.library.entity.info.RatingDetails

/**
 * @author Ruben Gees
 */
internal class FixRatingDetailsAdapter {

    private companion object {
        private const val INVALID_ARRAY = "[]"
    }

    @FromJson
    fun fromJson(jsonReader: JsonReader, delegate: JsonAdapter<RatingDetails>): RatingDetails? {
        val json = jsonReader.nextString()

        return when {
            json.isBlank() || json == INVALID_ARRAY -> RatingDetails(0, 0, 0, 0, 0)
            else -> delegate.fromJson(json)
        }
    }
}
